src/Language/Futhark/TypeChecker/Terms.hs

```diff
@@ -213,7 +213,7 @@ checkCoerce loc te e = do
   te_t_nonrigid <- makeNonExtFresh ext te_t
 
   onFailure (CheckingAscription te_t e_t) $
-    unify (mkUsage loc "size coercion") e_t te_t_nonrigid
+    unify (mkUsage loc "size coercion") te_t_nonrigid e_t
```

Comment:
Is this just for better error messages, or does it actually matter?  Unification is supposed to be commutative.

Reply:
It does matter, unification is not commutative when two free variables are to be unified: the first one is replaced by the second one, which can cause problems to `checkDoLoop` if a variable of the merge pattern is just coerced

Reply:
Do we have a test that shows this?  This could be a bug in unification.  When we have a choice of which free variable to pick, maybe we should pick the one that is bound at the outermost level.

Reply:
if reversing if commit, `tests/loops/pow2reduce.fut` and `tests/issue1424.fut` fails at typing.

For `tests/loops/pow2reduce.fut` (the one I studied to solve the problem):
- the merge type, which is the type of initial value with all sizes being freed, is `[loop₄]i32`
- `xs :> ...`: `[loop₄]i32` is unified with `[coerce₁₄]i32`
- when the pattern is `normTypeFully`, it takes the type `[coerce₁₄]i32`
- so it's not one of the variable introduced by the loop, and its variation is not detected

Reply:
I'll try to make this without this PR

Reply:
in master
```futhark
def main (xs : [10]i64) =
    loop xs for i < 10 do
        let n = 10 + i
        in (xs :> [n]i64) ++ [i]
```

current state, `futhark check` replies
```
Error at test.fut:2:10-11:
Initial loop values do not have expected type.
Expected: [k₈]i64
Actual:   [10]i64

Dimensions "k₈" and "10" do not match.

Note: "k₈" is unknown size returned by "++" at 4:12-32.
```

with this modification, it just accept